docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+
+### Algorithmic contracts
+Algorithmic contracts define how to process a given set of data structures to produce useful results. These are effectively code snippets which utilize the abstracted data structures provided by Data Structure Definition Contracts and Global Value Contract to produce useful output about a given program. Descriptions of these contracts may refer to functionality provided by other contracts to do their work. The algorithms provided in these contracts are designed to operate given the ability to read various primitive types and defined data structures from the process memory space, as well as perform general purpose computation.
+
+It is entirely reasonable for an algorithmic contract to have multiple entrypoints which take different inputs. For example imagine a contract which provides information about a `MethodTable`. It may provide the an api to get the `BaseSize` of a `MethodTable`, and an api to get the `DynamicTypeID` of a `MethodTable`. However, while the set of contracts which describe an older version of .NET may provide a means by which the `DynamicTypeID` may be acquired for a `MethodTable`, a newer runtime may not have that concept. In such a case, it is very reasonable to define that the `GetDynamicTypeID` api portion of that contract is defined to simply `throw new NotSupportedException();`
+
+For simplicity, as it can be expected that all developers who work on the .NET runtime understand C# to a fair degree, it is preferred that the algorithms be defined in C#, or at least psuedocode that looks like C#. It is also condsidered entirely permissable to refer to other specifications if the algorithm is a general purpose one which is well defined by the OS or some other body. (For example, it is expected that the unwinding algorithms will be defined by references into either the DWARF spec, or various Windows Unwind specifications.)
+
+For working with data from the target process/other contracts, the following C# interface is intended to be used within the algorithmic descriptions:
+
+``` csharp
+namespace DataContracts
+{
+    struct TargetPointer
+    {
+        public ulong Value;
+    }
+
+    enum FieldType
+    {
+        Int8Type,
+        UInt8Type,
+        Int16Type,
+        UInt16Type,
+        Int32Type,
+        UInt32Type,
+        Int64Type,
+        UInt64Type,
+        PointerType,
+
+        // Other values are dynamically assigned by the type definition rules
+    }
+
+    struct FieldLayout
+    {
+        public uint Offset;
+        public FieldType Type;
+    }
+
+    interface IAlgorithmContract
+    {
+        void Init();
+    }
+
+    interface IContract
+    {
+        string Name { get; }
+        uint Version { get; }
```

Comment:
as I mentioned above, I think this should also include `IReadOnlyList<uint> Subsumes { get; }` to list the older versions of a contract that a reader for a newer contract can understand.